Documentação do código (#6)
ThundeRatz/STM32RF24: CONTRIBUTING.md

# Contributing to STM32RF24 🚀

Para a versão em PT-BR 🇧🇷 desse documento, [veja aqui](CONTRIBUTING.pt-br.md).

First of all, thanks for your interest! We really appreciate the participation of the community! ✨

> Here we describe a set of guidelines when contributing to the VSS simulation. These are not strict rules, so use your best judgment, and in case you have questions, contact us on comp@thunderatz.org.

Reviewer: ```suggestion
Here we describe a set of guidelines when contributing to STM32RF24. These are not strict rules, so use your best judgment, and in case you have questions, contact us on comp@thunderatz.org.
```

Reviewer 2: Pegou no pulo

#### Reporting Bugs 🐛

- Use a clear title
- Specify the version of the library
- Specify OS, VM (if applicable), project's microcontroller, and other configurations that may be usefull
- Describe the steps to reproduce the bug
- Describe the observed and the expected behavior
- Include screenshots, gifs, any type of reference that helps to explain the problem

#### Suggesting Enhancements ✨

- Use a clear title
- Describe the suggestion step-by-step
- Describe the expected behavior after implementing this idea
- Explain why this new feature or update can be usefull

#### Imporoving documentation 📝

- Use a clear title
- Specify the files that need to be documented
- Explain your suggestion and why would it be better/clearer

### Pull Requests

If you want to contribute with code to the project, search for an open issue and start developing. When you are ready, open a Pull Request and we will review it.

A few recomendations:

- Describe exactly what you've done, always as clear as possible
- Link the applicable issue in your Pull Request (if there is none, please, create a new one)
- Make sure that you are following the [Syleguide 💄](#styleguide-)

## Styleguide 💄

> ### Formatação do código

Reviewer: ```suggestion
### Code formatting
```

The code is formatted following specific rules, for more information on how to ensure that your code is formatted correctly, see the [🎨 Formatting](./README.md#-formatting) section on the README.

### Coding in C

There are some guidelines for organizing C code that should be followed.

> "Programs must be written for people to read,
> and only incidentally for machines to execute."

― Harold Abelson, Structure and Interpretation of Computer Programs

#### Scopes

> Always reduce the scopes of the to the minimum possible, avoiding global, external and static variables, giving preference to local variables. The same is ture for functions, do not create public functions if they are only to be used within a file, leave them private.

Reviewer: O escopo do que?

Author: Justo, bom explicitar

Reviewer 2: Tá escrito "ture" em vez de "true" ai no meio

For variables or constants, for example, that have a unit of measure, include this unit of measure in the name, this makes it easier to understand its meaning.

In order to indicate that a variable is a pointer, add `p_` to the beginning of the variable name, if it is a pointer to a pointer, for example, add `pp_`.

Global variables in a file and external variables should be avoided, especially external ones, but if they are really necessary, they should be identified to facilitate the understanding of isolated parts of the code. Add `m_` to the beginning of a variable name to indicate that it is a global variable in a file. And to indicate that a variable is a global external, add `g_` at the beginning of its name. If you had an external pointer, for example, you would use `gp_`.

For more naming tips, see the article ["How To Create Meaningful Names In Code"](https://medium.com/better-programming/how-to-create-meaningful-names-in-code-20d7476537d4).

#### Variable types

Make coherent choices for your variable types and your functions return types, besides its parameters, for examples, do not use `float` if your variable is a unit counter.

> Make a consistent choice for the types of variables and types of return of functions, in addition to their parameters, for example, do not use `float` if your variable will be a unit counter.

Reviewer: ```suggestion
Make coherent choices for your variable types and your functions return and parameters types. For examples, do not use `float` if your variable is a unit counter.
```

> Além disso, para variáveis inteiras faça distinção entre `signed` e `unsigned` e especifique seu tamanho em bits, isso garante mais controle sobre os valores que cada variável pode ter, como também otimiza a ocupação da memória. Para isso utilize a biblioteca `stdint.h`, podendo assim, por exemplo, se utilizar os tipos `uint8_t` para inteiros de 8 bits sem sinal e `int16_t` para inteiros de 16 bits com sinal.

Reviewer: Lembrar de apagar essa parte em português q sobrou aqui

Besides that, for integer variables, distinguish between `signed` and `unsigned` and specify their size in bits, this ensures more control over the values that each variable can have, as well as optimizes the memory occupation. To do this, use the `stdint.h` library, than, for example, you can use the types `uint8_t` for unsigned 8-bit integers and `int16_t` for signed 16-bit integers.

#### Magic numbers/Constants

> Avoid [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)) in your code, in addition to obscuring the understanding of it, they also make maintenance difficult. To solve this, use constants defined with `#define` at the beginning of your code.

Reviewer: ```suggestion
Avoid [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)) in your code. In addition to decreasing readability, they also make maintenance difficult. To solve this, use constants defined with `#define` at the beginning of your code.
```

Não sei, achei meio estranho aquela frase :p
Mas só uma sugestão kkkkk

― Harold Abelson, Structure and Interpretation of Computer Programs

#### Variables and functions scopes

> Always reduce the scopes of the to the minimum possible, avoiding global, external and static variables, giving preference to local variables. The same is ture for functions, do not create public functions if they are only to be used within a file, leave them private.

Reviewer: Quando eu falei do escopo tava querendo dizer no texto hahahha

```suggestion
Always reduce the scopes of the variables to the minimum possible, avoiding global, external and static, giving preference to local variables. The same is true for functions, do not create public functions if they are only to be used within a file, leave them private.
```

Author: Ahhh justo shuahsaushahsa
